metrics/meter.go

```diff
@@ -0,0 +1,44 @@
+package metrics
```

Comment:
Let's not move this library, it's not actually libp2p specific and people should be able to use it outside of libp2p. Really, I only put it in the libp2p org because I didn't want to leave it as a personal repo.

Reply:
The metrics interfaces reference libp2p types (peer.ID and protocol.ID). They aren’t used beyond libp2p or projects that transitively depend on libp2p: https://godoc.org/github.com/libp2p/go-libp2p-metrics?importers. I think it’s pretty tied to libp2p.

Reply:
I'm talking about the files taken from https://github.com/libp2p/go-flow-metrics.

Reply:
Ah, I see. I did check the importers of go-flow-metrics to determine whether to pull it in or not. [The grand total is 1: us](https://godoc.org/github.com/libp2p/go-flow-metrics?importers).

The library is not actively maintained (no significant commits since December 2017), so I don't think we'll be doing any damage if we pulled it into core.

Reply:
I don't think this library belongs to core either; let's not bloat.

Reply:
Besides, aren't we phasing it out in favour of opencensus?

Reply:
> Ah, I see. I did check the importers of go-flow-metrics to determine whether to pull it in or not. The grand total is 1: us.

I wrote it for us and haven't advertised it so that's not exactly surprising. However, if we pull it into core, nobody outside of libp2p will be able to use it without also pulling in libp2p.

Please _do not_ include this in the core bundle.

> The library is not actively maintained (no significant commits since December 2017), so I don't think we'll be doing any damage if we pulled it into core.

Three haven't been any significant commits since December 2017 because there haven't been any bugs since December 2017. It's completely finished and actively maintained by me (should someone find a bug).

Reply:
@vyzo go-libp2p-metrics and go-flow-metrics are two different libraries. We use this library for bandwidth metrics.

Reply:
@Stebalien gotcha, will revert `go-flow-metrics`. Sidenote: I re-read the conversation and I think I may have come off a bit abrupt. Apologies, it wasn't my intention.

Reply:
Revisiting this code to back out the change, I realise why I decided to coalesce both modules. go-libp2p-metrics _depends on_ go-flow-metrics. The `BandwidthCounter` type uses `flow.Meter` to reify the bookkeeping.

We have two options:

1. pull `Reporter` and its associated struct `Stats` into go-libp2p-core, leaving `BandwidthCounter` behind (in go-libp2p-metrics), considering it the _glue_ between the abstraction and the implementation using go-flow-metrics.
  * I dislike this option because we reduce module count by zero (concerning metrics), and the resulting modules are very slim by themselves.
2. pull all of go-libp2p-metrics in, and make all of go-libp2p-core depend on go-flow-metrics.

I'm choosing option 2 because we're moving to OC anyway. Shout if you disagree.

Reply:
> Sidenote: I re-read the conversation and I think I may have come off a bit abrupt. Apologies, it wasn't my intention.

Not at all, I was just afraid I wasn't making my position on this clear. I consider go-flow-metrics to be something like go-buffer-pool, useful outside of libp2p, so I really didn't want to see it absorbed.

> pull all of go-libp2p-metrics in, and make all of go-libp2p-core depend on go-flow-metrics.

SGTM. go-libp2p-metrics is very much a libp2p package. BandwidthCounter would, ideally, go in a separate package (maybe even in go-libp2p?) as it's an implementation of the `Reporter` interface but I don't feel strongly either way.

(I'm also not really a fan of the interface itself but we can revisit that in a future refactor)